connectors/utils.py

```diff
@@ -361,18 +360,28 @@ async def put(self, coroutine, result_callback=None):
 
         If provided, `result_callback` will be called when the task is done.
         """
-        # If self.tasks has reached its max size, we wait for one task to finish
-        if len(self.tasks) >= self.max_concurrency:
-            await self._task_over.wait()
-            # rearm
-            self._task_over.clear()
+        await self._sem.acquire()
         task = asyncio.create_task(coroutine())
         self.tasks.append(task)
         task.add_done_callback(
             functools.partial(self._callback, result_callback=result_callback)
         )
         return task
 
+    async def try_put(self, coroutine, result_callback=None):
+        """Tries to add a coroutine for immediate execution.
+
+        If the number of running tasks reach `max_concurrency`, this
+        function return a None task immediately
+
+        If provided, `result_callback` will be called when the task is done.
+        """
+
+        if self._sem.locked():
```

Comment:
Should we add a debug log line here? Could be interesting for support as there's a potential action attached to it (increasing/decreasing `max_concurrency` config). We could also add a debug log line further up in the call stack and make it specific for the sync job (or whoever the caller is).

Reply:
> Could be interesting for support as there's a potential action attached to it (increasing/decreasing `max_concurrency` config).

This won't take effect until you restart the service.

If `self._sem.locked()` returns `True`, it means the task queue is full: https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore.locked

Reply:
> This won't take effect until you restart the service.
It still could be something you want to notice as you can do something about it (restarting and increasing `max_concurrency`), right? Just that it's more obvious, if the service starts to become slower than expected due to reaching the `max_concurrency` limit. Or do we log it somewhere else in a different form?

Reply:
>Or do we log it somewhere else in a different form?

I'd like to log it in the caller of this method, as `ConcurrentTasks` doesn't have the context.

E.g. We can log in `JobExecutionService`:
> The Sync job 123 is skipped and will retry in 30 seconds because the task pool is full.